### PR TITLE
[MRG+1] Remove __str__ and __repr__ from settings, introduce copy_to_dict()

### LINKED ISSUE #1732: BaseSettings.__repr__ is too verbose for scrapy shell
Settings object got `__repr__` in https://github.com/scrapy/scrapy/pull/1149, but it looks weird in `scrapy shell`:

> [s] Available Scrapy objects:
[s]   crawler    &lt;scrapy.crawler.Crawler object at 0x10c042fd0>
[s]   item       {}
[s]   request    &lt;GET http://yellowpages.co.th>
[s]   response   &lt;302 http://yellowpages.co.th>
[s]   settings   {'CLOSESPIDER_ITEMCOUNT': &lt;SettingsAttribute value=0 priority=0>, 'RETRY_HTTP_CODES': &lt;SettingsAttribute value=[500, 502, 503, 504, 408] priority=0>, 'HTTPCACHE_DIR': &lt;SettingsAttribute value='httpcache-2' priority=20>, 'MEMUSAGE_REPORT': &lt;SettingsAttribute value=False priority=0>, 'DOWNLOAD_HANDLERS_BASE': &lt;SettingsAttribute value=&lt;BaseSettings {'s3': &lt;SettingsAttribute value='scrapy.core.downloader.handlers.s3.S3DownloadHandler' priority=0>, 'ftp': &lt;SettingsAttribute value='scrapy.core.downloader.handlers.ftp.FTPDownloadHandler' priority=0>, 'http': &lt;SettingsAttribute value='scrapy.core.downloader.handlers.http.HTTPDownloadHandler' priority=0>, 'https': &lt;SettingsAttribute value='scrapy.core.downloader.handlers.http.HTTPDownloadHandler' priority=0>, 'file': &lt;SettingsAttribute value='scrapy.core.downloader.handlers.file.FileDownloadHandler' priority=0>}> priority=0>, 'RETRY_PRIORITY_ADJUST': &lt;SettingsAttribute value=-1 priority=0>, 'MAIL_FROM': &lt;SettingsAttribute value='scrapy@localhost' priority=0>, 'HTTPCACHE_EXPIRATION_SECS': &lt;SettingsAttribute value=86400 priority=20>, 'SPIDER_LOADER_CLASS': &lt;SettingsAttribute value='scrapy.spiderloader.SpiderLoader' priority=0>, 'COMPRESSION_ENABLED': &lt;SettingsAttribute value=True priority=0>, 'DOWNLOAD_TIMEOUT': &lt;SettingsAttribute value=180 priority=0>, 'MAIL_PASS': &lt;SettingsAttribute value=None priority=0>, 'MEMUSAGE_LIMIT_MB': &lt;SettingsAttribute value=0 priority=0>, 'EXTENSIONS': &lt;SettingsAttribute value=&lt;BaseSettings {}> priority=0>, 'DEPTH_PRIORITY': &lt;SettingsAttribute value=0 priority=0>, 'TELNETCONSOLE_HOST': &lt;SettingsAttribute value='127.0.0.1' priority=0>, 'MEMDEBUG_NOTIFY': &lt;SettingsAttribute value=[] priority=0>, 'HTTPPROXY_AUTH_ENCODING': &lt;SettingsAttribute value='latin-1' priority=0>, 'DOWNLOAD_WARNSIZE': &lt;SettingsAttribute value=33554432 priority=0>, 'SPIDER_MODULES': &lt;SettingsAttribute value=['acrawler.spiders'] priority=20>, 'RETRY_TIMES': &lt;SettingsAttribute value=2 priority=0>, 'TELNETCONSOLE_PORT': &lt;SettingsAttribute value=[6023, 6073] priority=0>, 'TELNETCONSOLE_ENABLED': &lt;SettingsAttribute value=False priority=20>, 'DOWNLOADER_MIDDLEWARES': &lt;SettingsAttribute value=&lt;BaseSettings {}> priority=0>, 'HTTPCACHE_DBM_MODULE': &lt;SettingsAttribute value='anydbm' priority=0>, 'ROBOTSTXT_OBEY': &lt;SettingsAttribute value=True priority=20>, 'DEPTH_LIMIT': &lt;SettingsAttribute value=0 priority=0>, 'REACTOR_THREADPOOL_MAXSIZE': &lt;SettingsAttribute value=10 priority=0>, 'FEED_EXPORT_FIELDS': &lt;SettingsAttribute value=None priority=0>, 'CLOSESPIDER_PAGECOUNT': &lt;SettingsAttribute value=0 priority=0>, 'LOG_SHORT_NAMES': &lt;SettingsAttribute value=False priority=0>, 'AUTOTHROTTLE_MAX_DELAY': &lt;SettingsAttribute value=60 priority=20>, 'URLLENGTH_LIMIT': &lt;SettingsAttribute value=2083 priority=0>, 'FEED_EXPORTERS': &lt;SettingsAttribute value=&lt;BaseSettings {}> priority=0>, 'LOG_ENCODING': &lt;SettingsAttribute value='utf-8' priority=0>, 'FEED_EXPORTERS_BASE': &lt;SettingsAttribute value=&lt;BaseSettings {'xml': &lt;SettingsAttribute value='scrapy.exporters.XmlItemExporter' priority=0>, 'jsonlines': &lt;SettingsAttribute value='scrapy.exporters.JsonLinesItemExporter' priority=0>, 'jl': &lt;SettingsAttribute value='scrapy.exporters.JsonLinesItemExporter' priority=0>, 'json': &lt;SettingsAttribute value='scrapy.exporters.JsonItemExporter' priority=0>, 'csv': &lt;SettingsAttribute value='scrapy.exporters.CsvItemExporter' priority=0>, 'pickle': &lt;SettingsAttribute value='scrapy.exporters.PickleItemExporter' priority=0>, 'marshal': &lt;SettingsAttribute value='scrapy.exporters.MarshalItemExporter' priority=0>}> priority=0>, 'FEED_FORMAT': &lt;SettingsAttribute value='jsonlines' priority=0>, 'DOWNLOAD_DELAY': &lt;SettingsAttribute value=0 priority=0>, 'HTTPCACHE_GZIP': &lt;SettingsAttribute value=False priority=0>, 'DOWNLOADER_MIDDLEWARES_BASE': &lt;SettingsAttribute value=&lt;BaseSettings {'scrapy.downloadermiddlewares.useragent.UserAgentMiddleware': &lt;SettingsAttribute value=400 priority=0>, 'scrapy.downloadermiddlewares.defaultheaders.DefaultHeadersMiddleware': &lt;SettingsAttribute value=550 priority=0>, 'scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware': &lt;SettingsAttribute value=590 priority=0>, 'scrapy.downloadermiddlewares.chunked.ChunkedTransferMiddleware': &lt;SettingsAttribute value=830 priority=0>, 'scrapy.downloadermiddlewares.robotstxt.RobotsTxtMiddleware': &lt;SettingsAttribute value=100 priority=0>, 'scrapy.downloadermiddlewares.downloadtimeout.DownloadTimeoutMiddleware': &lt;SettingsAttribute value=350 priority=0>, 'scrapy.downloadermiddlewares.redirect.RedirectMiddleware': &lt;SettingsAttribute value=600 priority=0>, 'scrapy.downloadermiddlewares.ajaxcrawl.AjaxCrawlMiddleware': &lt;SettingsAttribute value=560 priority=0>, 'scrapy.downloadermiddlewares.httpproxy.HttpProxyMiddleware': &lt;SettingsAttribute value=750 priority=0>, 'scrapy.downloadermiddlewares.httpcache.HttpCacheMiddleware': &lt;SettingsAttribute value=900 priority=0>, 'scrapy.downloadermiddlewares.httpauth.HttpAuthMiddleware': &lt;SettingsAttribute value=300 priority=0>, 'scrapy.downloadermiddlewares.retry.RetryMiddleware': &lt;SettingsAttribute value=500 priority=0>, 'scrapy.downloadermiddlewares.stats.DownloaderStats': &lt;SettingsAttribute value=850 priority=0>, 'scrapy.downloadermiddlewares.cookies.CookiesMiddleware': &lt;SettingsAttribute value=700 priority=0>, 'scrapy.downloadermiddlewares.redirect.MetaRefreshMiddleware': &lt;SettingsAttribute value=580 priority=0>}> priority=0>, 'DNSCACHE_ENABLED': &lt;SettingsAttribute value=True priority=0>, 'CONCURRENT_REQUESTS_PER_IP': &lt;SettingsAttribute value=0 priority=0>, 'EDITOR': &lt;SettingsAttribute value='nano' priority=0>, 'MAIL_HOST': &lt;SettingsAttribute value='localhost' priority=0>, 'CONCURRENT_REQUESTS': &lt;SettingsAttribute value=100 priority=20>, 'AUTOTHROTTLE_START_DELAY': &lt;SettingsAttribute value=1 priority=20>, 'CLOSESPIDER_ERRORCOUNT': &lt;SettingsAttribute value=0 priority=0>, 'STATS_CLASS': &lt;SettingsAttribute value='scrapy.statscollectors.MemoryStatsCollector' priority=0>, 'FEED_STORAGES_BASE': &lt;SettingsAttribute value=&lt;BaseSettings {'': &lt;SettingsAttribute value='scrapy.extensions.feedexport.FileFeedStorage' priority=0>, 's3': &lt;SettingsAttribute value='scrapy.extensions.feedexport.S3FeedStorage' priority=0>, 'ftp': &lt;SettingsAttribute value='scrapy.extensions.feedexport.FTPFeedStorage' priority=0>, 'file': &lt;SettingsAttribute value='scrapy.extensions.feedexport.FileFeedStorage' priority=0>, 'stdout': &lt;SettingsAttribute value='scrapy.extensions.feedexport.StdoutFeedStorage' priority=0>}> priority=0>, 'REDIRECT_ENABLED': &lt;SettingsAttribute value=True priority=0>, 'AUTOTHROTTLE_ENABLED': &lt;SettingsAttribute value=True priority=20>, 'COMMANDS_MODULE': &lt;SettingsAttribute value='' priority=0>, 'AUTOTHROTTLE_DEBUG': &lt;SettingsAttribute value=False priority=0>, 'NEWSPIDER_MODULE': &lt;SettingsAttribute value='acrawler.spiders' priority=20>, 'LOG_UNSERIALIZABLE_REQUESTS': &lt;SettingsAttribute value=False priority=0>, 'DOWNLOAD_MAXSIZE': &lt;SettingsAttribute value=1073741824 priority=0>, 'MAIL_PORT': &lt;SettingsAttribute value=25 priority=0>, 'REFERER_ENABLED': &lt;SettingsAttribute value=True priority=0>, 'HTTPCACHE_POLICY': &lt;SettingsAttribute value='scrapy.extensions.httpcache.DummyPolicy' priority=0>, 'STATS_DUMP': &lt;SettingsAttribute value=True priority=0>, 'MEMUSAGE_NOTIFY_MAIL': &lt;SettingsAttribute value=[] priority=0>, 'DOWNLOAD_HANDLERS': &lt;SettingsAttribute value=&lt;BaseSettings {}> priority=0>, 'LOG_DATEFORMAT': &lt;SettingsAttribute value='%Y-%m-%d %H:%M:%S' priority=0>, 'LOG_LEVEL': &lt;SettingsAttribute value='DEBUG' priority=0>, 'DOWNLOADER_HTTPCLIENTFACTORY': &lt;SettingsAttribute value='scrapy.core.downloader.webclient.ScrapyHTTPClientFactory' priority=0>, 'REDIRECT_MAX_TIMES': &lt;SettingsAttribute value=20 priority=0>, 'REDIRECT_PRIORITY_ADJUST': &lt;SettingsAttribute value=2 priority=0>, 'DUPEFILTER_CLASS': &lt;SettingsAttribute value='scrapy.dupefilters.BaseDupeFilter' priority=10>, 'RETRY_ENABLED': &lt;SettingsAttribute value=True priority=0>, 'SPIDER_CONTRACTS': &lt;SettingsAttribute value=&lt;BaseSettings {}> priority=0>, 'HTTPCACHE_ENABLED': &lt;SettingsAttribute value=True priority=20>, 'LOG_ENABLED': &lt;SettingsAttribute value=True priority=0>, 'MAIL_USER': &lt;SettingsAttribute value=None priority=0>, 'HTTPCACHE_ALWAYS_STORE': &lt;SettingsAttribute value=False priority=0>, 'LOGSTATS_INTERVAL': &lt;SettingsAttribute value=0 priority=10>, 'DEFAULT_ITEM_CLASS': &lt;SettingsAttribute value='scrapy.item.Item' priority=0>, 'DNS_TIMEOUT': &lt;SettingsAttribute value=60 priority=0>, 'DEPTH_STATS': &lt;SettingsAttribute value=True priority=0>, 'DOWNLOADER_CLIENTCONTEXTFACTORY': &lt;SettingsAttribute value='scrapy.core.downloader.contextfactory.ScrapyClientContextFactory' priority=0>, 'MEMUSAGE_CHECK_INTERVAL_SECONDS': &lt;SettingsAttribute value=60.0 priority=0>, 'EXTENSIONS_BASE': &lt;SettingsAttribute value=&lt;BaseSettings {'scrapy.extensions.corestats.CoreStats': &lt;SettingsAttribute value=0 priority=0>, 'scrapy.extensions.feedexport.FeedExporter': &lt;SettingsAttribute value=0 priority=0>, 'scrapy.extensions.memdebug.MemoryDebugger': &lt;SettingsAttribute value=0 priority=0>, 'scrapy.extensions.memusage.MemoryUsage': &lt;SettingsAttribute value=0 priority=0>, 'scrapy.extensions.logstats.LogStats': &lt;SettingsAttribute value=0 priority=0>, 'scrapy.extensions.telnet.TelnetConsole': &lt;SettingsAttribute value=0 priority=0>, 'scrapy.extensions.closespider.CloseSpider': &lt;SettingsAttribute value=0 priority=0>, 'scrapy.extensions.spiderstate.SpiderState': &lt;SettingsAttribute value=0 priority=0>, 'scrapy.extensions.throttle.AutoThrottle': &lt;SettingsAttribute value=0 priority=0>}> priority=0>, 'FEED_STORAGES': &lt;SettingsAttribute value=&lt;BaseSettings {}> priority=0>, 'BOT_NAME': &lt;SettingsAttribute value='acrawler' priority=20>, 'SPIDER_CONTRACTS_BASE': &lt;SettingsAttribute value=&lt;BaseSettings {'scrapy.contracts.default.ScrapesContract': &lt;SettingsAttribute value=3 priority=0>, 'scrapy.contracts.default.UrlContract': &lt;SettingsAttribute value=1 priority=0>, 'scrapy.contracts.default.ReturnsContract': &lt;SettingsAttribute value=2 priority=0>}> priority=0>, 'METAREFRESH_MAXDELAY': &lt;SettingsAttribute value=100 priority=0>, 'CONCURRENT_REQUESTS_PER_DOMAIN': &lt;SettingsAttribute value=8 priority=0>, 'HTTPCACHE_IGNORE_HTTP_CODES': &lt;SettingsAttribute value=[] priority=0>, 'KEEP_ALIVE': &lt;SettingsAttribute value=True priority=10>, 'ITEM_PROCESSOR': &lt;SettingsAttribute value='scrapy.pipelines.ItemPipelineManager' priority=0>, 'MEMUSAGE_WARNING_MB': &lt;SettingsAttribute value=0 priority=0>, 'FEED_STORE_EMPTY': &lt;SettingsAttribute value=False priority=0>, 'COOKIES_DEBUG': &lt;SettingsAttribute value=False priority=0>, 'FEED_URI': &lt;SettingsAttribute value=None priority=0>, 'SPIDER_MIDDLEWARES': &lt;SettingsAttribute value=&lt;BaseSettings {}> priority=0>, 'DOWNLOADER': &lt;SettingsAttribute value='scrapy.core.downloader.Downloader' priority=0>, 'AUTOTHROTTLE_TARGET_CONCURRENCY': &lt;SettingsAttribute value=1.0 priority=20>, 'USER_AGENT': &lt;SettingsAttribute value='acrawler' priority=20>, 'AJAXCRAWL_ENABLED': &lt;SettingsAttribute value=False priority=0>, 'COOKIES_ENABLED': &lt;SettingsAttribute value=False priority=20>, 'DNSCACHE_SIZE': &lt;SettingsAttribute value=10000 priority=0>, 'LOG_FORMAT': &lt;SettingsAttribute value='%(asctime)s [%(name)s] %(levelname)s: %(message)s' priority=0>, 'ITEM_PIPELINES': &lt;SettingsAttribute value=&lt;BaseSettings {}> priority=0>, 'LOG_FORMATTER': &lt;SettingsAttribute value='scrapy.logformatter.LogFormatter' priority=0>, 'HTTPCACHE_IGNORE_RESPONSE_CACHE_CONTROLS': &lt;SettingsAttribute value=[] priority=0>, 'METAREFRESH_ENABLED': &lt;SettingsAttribute value=True priority=0>, 'HTTPCACHE_IGNORE_MISSING': &lt;SettingsAttribute value=False priority=0>, 'HTTPCACHE_IGNORE_SCHEMES': &lt;SettingsAttribute value=['file'] priority=0>, 'SCHEDULER_MEMORY_QUEUE': &lt;SettingsAttribute value='scrapy.squeues.LifoMemoryQueue' priority=0>, 'SCHEDULER_DISK_QUEUE': &lt;SettingsAttribute value='scrapy.squeues.PickleLifoDiskQueue' priority=0>, 'RANDOMIZE_DOWNLOAD_DELAY': &lt;SettingsAttribute value=True priority=0>, 'SETTINGS_MODULE': &lt;SettingsAttribute value='acrawler.settings' priority=20>, 'TEMPLATES_DIR': &lt;SettingsAttribute value='/Users/kmike/svn/scrapy/scrapy/templates' priority=0>, 'LOG_STDOUT': &lt;SettingsAttribute value=False priority=0>, 'CONCURRENT_ITEMS': &lt;SettingsAttribute value=100 priority=0>, 'DOWNLOADER_STATS': &lt;SettingsAttribute value=True priority=0>, 'LOG_FILE': &lt;SettingsAttribute value=None priority=0>, 'HTTPCACHE_STORAGE': &lt;SettingsAttribute value='scrapy.extensions.httpcache.FilesystemCacheStorage' priority=20>, 'MEMDEBUG_ENABLED': &lt;SettingsAttribute value=False priority=0>, 'FEED_URI_PARAMS': &lt;SettingsAttribute value=None priority=0>, 'DEFAULT_REQUEST_HEADERS': &lt;SettingsAttribute value=&lt;BaseSettings {'Accept-Language': &lt;SettingsAttribute value='en' priority=0>, 'Accept': &lt;SettingsAttribute value='text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8' priority=0>}> priority=0>, 'CLOSESPIDER_TIMEOUT': &lt;SettingsAttribute value=0 priority=0>, 'SCHEDULER': &lt;SettingsAttribute value='scrapy.core.scheduler.Scheduler' priority=0>, 'SPIDER_MIDDLEWARES_BASE': &lt;SettingsAttribute value=&lt;BaseSettings {'scrapy.spidermiddlewares.referer.RefererMiddleware': &lt;SettingsAttribute value=700 priority=0>, 'scrapy.spidermiddlewares.httperror.HttpErrorMiddleware': &lt;SettingsAttribute value=50 priority=0>, 'scrapy.spidermiddlewares.depth.DepthMiddleware': &lt;SettingsAttribute value=900 priority=0>, 'scrapy.spidermiddlewares.urllength.UrlLengthMiddleware': &lt;SettingsAttribute value=800 priority=0>, 'scrapy.spidermiddlewares.offsite.OffsiteMiddleware': &lt;SettingsAttribute value=500 priority=0>}> priority=0>, 'ITEM_PIPELINES_BASE': &lt;SettingsAttribute value=&lt;BaseSettings {}> priority=0>, 'STATSMAILER_RCPTS': &lt;SettingsAttribute value=[] priority=0>, 'MEMUSAGE_ENABLED': &lt;SettingsAttribute value=False priority=0>}
[s]   spider     &lt;DefaultSpider 'default' at 0x10ebdcf90>
[s] Useful shortcuts:
[s]   shelp()           Shell help (print this help)
[s]   fetch(req_or_url) Fetch request (or URL) and update local objects
[s]   view(response)    View response in a browser

### PR DESCRIPTION
**Pull request by redapple**
Settings instances as dict's are easier to print or pretty print in the shell

Fixes #1732

### CODE REVIEW

**Comment by redapple** [comment]
@kmike , would that work for you for #1732 ?

**Comment by redapple** [comment]
ok, this doesn't pass `scrapy settings` test
```
$ scrapy settings --get EXTENSIONS -s EXTENSIONS='{"tests.test_cmdline.extensions.DummyExtension": 100}'
<scrapy.settings.BaseSettings object at 0x7f63bf407350>
```

**Comment by Digenis** [comment]
Have you considered IPython's methods for pretty printing?

**Comment by redapple** [comment]
@Digenis , no, what do you suggest?

**Comment by Digenis** [comment]
http://ipython.readthedocs.org/en/stable/api/generated/IPython.lib.pretty.html?highlight=_repr_pretty_#extending
If the settings class defines `_repr_pretty_`, IPython will use it in its shell.
Implementing dict-alike formatting should be trivial.

**Comment by redapple** [comment]
@Digenis ,
this is the output I get: https://gist.github.com/redapple/ba015fab046310b49cdc
simplistic implementation here but works fine IMO, thanks for the tip!

File: scrapy/settings/__init__.py [hunk 1]
```diff
@@ -368,11 +369,31 @@ def __iter__(self):
...
+        Modifications to the returned dict won't be reflected on the original
+        settings.
+
+        This method can be useful for example for printing settings
+        in Scrapy shell.
+        """
+        settings = self.copy()
+        return settings._to_dict()
+
+    def _repr_pretty_(self, p, cycle):
```
Comment (eliasdorneles) [inline scrapy/settings/__init__.py]: uh, sorry for the dumb question, but where is this method being used?

Comment (redapple) [inline scrapy/settings/__init__.py]: there's no dumb question :)
it's used by ipython shell.
I see the effect of it being used in my local tests. I don't know how to test it in Travis

Comment (eliasdorneles) [inline scrapy/settings/__init__.py]: ahh got it, that's how you get that nice output! cool :)
perhaps worth a docstring?

Comment (redapple) [inline scrapy/settings/__init__.py]: Good idea. Will update

Comment (nramirezuy) [inline scrapy/settings/__init__.py]: don't you feel like returning a `dict` on `__repr__` is wrong ?

File: scrapy/settings/__init__.py [hunk 2]
```diff
@@ -368,11 +369,31 @@ def __iter__(self):
...
+        and their priorities as the current settings.
+
+        Modifications to the returned dict won't be reflected on the original
+        settings.
+
+        This method can be useful for example for printing settings
+        in Scrapy shell.
+        """
+        settings = self.copy()
+        return settings._to_dict()
```
Comment (nramirezuy) [inline scrapy/settings/__init__.py]: http://www.machinerytrader.com/listingsdetail/detail.aspx?OHID=11518535

Comment (redapple) [inline scrapy/settings/__init__.py]: @nramirezuy , `__repr__` is removed, `copy_to_dict` returns a dict

File: scrapy/settings/__init__.py [hunk 3]
```diff
@@ -368,11 +369,31 @@ def __iter__(self):
     def __len__(self):
         return len(self.attributes)
 
-    def __str__(self):
-        return str(self.attributes)
```
Comment (eliasdorneles) [inline scrapy/settings/__init__.py]: how about keeping `__str__` with:

    def __str__(self):
        return str(self._to_dict())

File: scrapy/settings/__init__.py (hunk 2 above)
Comment (nramirezuy) [inline scrapy/settings/__init__.py]: yes, sorry.

**Comment by nramirezuy** [comment]
I just find weird the removal of `__repr__` but the addition of a `__repr__` that only "works" on IPython.

File: scrapy/settings/__init__.py (hunk 3 above)
Comment (redapple) [inline scrapy/settings/__init__.py]: current shell implementation uses [`b.append("  %-10s %s" % (k, v))`](https://github.com/scrapy/scrapy/blob/master/scrapy/shell.py#L141), so `__str__` as a dict would be the same as in #1732

**Comment by redapple** [comment]
@nramirezuy , that's the best I've got to without changing too much

**Comment by redapple** [comment]
@nramirezuy , `_repr_pretty_` was added after https://github.com/scrapy/scrapy/pull/1746#issuecomment-177164854
it can be removed.

File: scrapy/settings/__init__.py (hunk 3 above)
Comment (eliasdorneles) [inline scrapy/settings/__init__.py]: hm, perhaps the shell should use `%r` instead (which would call repr)?

**Comment by nramirezuy** [comment]
@redapple Is not a problem, just feels weird;  like I would expect `__repr__` to be implemented too :smile:

**Pull request merged**